### PR TITLE
More annotations for `synapse.logging`, part 1

### PR DESCRIPTION
**Pull request by DMRobertson**
Goal is to make `synapse.logging` pass `no-untyped-defs`. This is a first step.

Some work also done in #10943 in this area.

### CODE REVIEW

File: synapse/logging/context.py [hunk 1]
```diff
@@ -711,16 +735,19 @@ def nested_logging_context(suffix: str) -> LoggingContext:
     )
 
 
-def preserve_fn(f):
+def preserve_fn(f: F) -> F:
```
Comment (squahtx) [inline synapse/logging/context.py:738]: We might want add a comment noting that the return type of `F` changes into a `Deferred[T]` if the original return type was not a `Deferred`.

I don't think we came to a consensus in #synapse-devs on whether it was okay to use slightly misleading type hints in this kind of situation.

Comment (DMRobertson) [reply synapse/logging/context.py:738]: Ahh, I didn't spot that this also happens in `run_in_background`. Hmm, maybe this needs more thought?

**Comment by clokep** [comment]
@DMRobertson I think *most* of this PR was done in #11556. Should we close this for now?

**Comment by DMRobertson** [comment]
Agreed!

**Pull request closed without merging**